libsolidity/analysis/ContractLevelChecker.cpp

```diff
@@ -666,56 +666,134 @@ void ContractLevelChecker::checkBaseABICompatibility(ContractDefinition const& _
+				nodes[_function] = currentNode;
+				nodeInv[currentNode] = _function;
+				if (_function->overrides())
+					for (auto const* baseFunction: _function->annotation().baseFunctions)
+						addEdge(currentNode, visit(baseFunction));
+				else
+					addEdge(currentNode, 1);
+
+				return currentNode;
+			}
+		} overrideGraph(baseFunctions);
 
-		// Function has been explicitly overridden
-		if (contains_if(
-			contractFuncs,
-			[&] (FunctionDefinition const* _f) {
-				return hasEqualNameAndParameters(*_f, **it);
+		// Detect cut vertices following https://en.wikipedia.org/wiki/Biconnected_component#Pseudocode
+		// Can ignore the root node, since it is never a cut vertex in our case.
+		struct CutVertexFinder
+		{
+			CutVertexFinder(OverrideGraph const& _graph): m_graph(_graph)
+			{
+				run();
 			}
-		))
-			continue;
+			std::set<FunctionDefinition const*> const& cutVertices() const { return m_cutVertices; }
 
-		set<FunctionDefinition const*> ambiguousFunctions;
-		SecondarySourceLocation ssl;
+		private:
+			OverrideGraph const& m_graph;
+
+			std::vector<bool> m_visited = std::vector<bool>(m_graph.numNodes, false);
+			std::vector<int> m_depths = std::vector<int>(m_graph.numNodes, -1);
+			std::vector<int> m_low = std::vector<int>(m_graph.numNodes, -1);
+			std::vector<int> m_parent = std::vector<int>(m_graph.numNodes, -1);
+			std::set<FunctionDefinition const*> m_cutVertices{};
+
+			void run(int _u = 0, int _depth = 0)
+			{
+				m_visited.at(_u) = true;
+				m_depths.at(_u) = m_low.at(_u) = _depth;
+				for (int v: m_graph.edges.at(_u))
+				{
+					if (!m_visited.at(v))
+					{
+						m_parent[v] = _u;
+						run(v, _depth + 1);
+						if (m_low[v] >= m_depths[_u] && m_parent[_u] != -1)
+							m_cutVertices.insert(m_graph.nodeInv.at(_u));
+						m_low[_u] = min(m_low[_u], m_low[v]);
+					}
+					else if (v != m_parent[_u])
+						m_low[_u] = min(m_low[_u], m_depths[v]);
+				}
+			}
+		} cutVertexFinder{overrideGraph};
 
-		for (;begin != end; begin++)
+		// Remove all base functions overridden by cut vertices (they don't need to be overridden).
+		for (auto const* function: cutVertexFinder.cutVertices())
 		{
-			ambiguousFunctions.insert(*begin);
-			ssl.append("Definition here: ", (*begin)->location());
+			std::set<FunctionDefinition const*> toTraverse = function->annotation().baseFunctions;
+			while (!toTraverse.empty())
+			{
+				auto const *base = *toTraverse.begin();
+				toTraverse.erase(toTraverse.begin());
+				baseFunctions.erase(base);
+				for (auto const* f: base->annotation().baseFunctions)
+					toTraverse.insert(f);
+			}
+			// Remove unimplemented base functions at the cut vertices itself as well.
+			if (!function->isImplemented())
+				baseFunctions.erase(function);
 		}
 
-		// Make sure the functions are not from the same base contract
-		if (ambiguousFunctions.size() == 1)
+		// If more than one function is left, they have to be overridden.
+		if (baseFunctions.size() <= 1)
```

Comment:
As @leonardoalt noted: to actually conform to the last bit of the spec this probably needs to be:
```
if (baseFunctions.empty() || (baseFunctions.size() == 1 && (*baseFunctions.begin())->isImplemented()))
    continue;
```
However, for the "only one unimplemented" function we need to be careful about the error message, since the conflicting function in the cut vertex was removed in the loop above...
And it might be worth to think about whether this part of the spec makes sense.

Reply:
This hinges solely on the question whether this is valid:
```
abstract contract I { function f() external; }
abstract contract A is I {}
abstract contract B is I { function f() external override; }
abstract contract C is A, B {}
```
If this can be extended to an actually weird case, we need to change the code - if there is no weird case connected to this, we can drop the last bit of the spec.

Note that if ``f`` is implemented in ``I``, then not implementing it in ``B`` is always an independent error (since implemented functions can only be overridden by implemented functions).

Reply:
@chriseth and I talked, we think it should be valid and the spec should be fixed